### PR TITLE
Log unhandled exceptions thrown from the AppFunc

### PR DESCRIPTION
**Pull request by halter73**

### CODE REVIEW

**Comment by halter73** [comment]
@Tratcher @moozzyk

**Comment by Tratcher** [comment]
Do you want to check these while you're at it? https://github.com/aspnet/KestrelHttpServer/issues/129

**Comment by halter73** [comment]
@Tratcher @davidfowl @lodejard With this PR, Kestrel should no log all application errors and keep running where possible addressing #129.

I also did a little cleanup of our connection shutdown logic in Frame.ProduceEnd, since we gracefully shutdown the connection in Frame.RequestProcessingAsync.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 1]
```diff
@@ -261,12 +262,19 @@ private void FireOnStarting()
             {
                 foreach (var entry in onStarting)
                 {
-                    entry.Key.Invoke(entry.Value).Wait();
+                    try
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: This try/catch should be outside the foreach, we don't want to keep executing the OnStarted callbacks if the request has failed. OnCompleted is different.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 2]
```diff
@@ -411,28 +413,43 @@ public void ProduceContinue()
...
-        public void ProduceStart(bool immediate = true, bool appCompleted = false)
+        public async Task ProduceStartAndFireOnStarting(bool immediate = true)
+        {
+            if (_responseStarted) return;
+
+            await FireOnStarting();
+           
+            if (_applicationFailed) 
+            {
+                throw new ObjectDisposedException(typeof(Frame).FullName);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: We should at least include a message saying that the response was aborted due to an application exception. We could even include the exception as a nested exception.

**Comment by Tratcher** [comment]
:watch:

**Comment by halter73** [comment]
@Tratcher good?

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**